daft/runners/ray_runner.py

```diff
@@ -655,6 +656,141 @@ def teardown_actor_pool(self, name: str) -> None:
+            elif len(next_step.instructions) == 0:
+                logger.debug("Running task synchronously in main thread: %s", next_step)
+                assert (
+                    len(next_step.partial_metadatas) == 1
+                ), "No-op tasks must have one output by definition, since there are no instructions to run"
+                [single_partial] = next_step.partial_metadatas
+                if single_partial.num_rows is None:
+                    [single_meta] = ray.get(get_metas.remote(next_step.inputs))
+                    accessor = PartitionMetadataAccessor.from_metadata_list(
+                        [single_meta.merge_with_partial(single_partial)]
+                    )
+                else:
+                    accessor = PartitionMetadataAccessor.from_metadata_list(
+                        [
+                            PartitionMetadata(
+                                num_rows=single_partial.num_rows,
+                                size_bytes=single_partial.size_bytes,
+                                boundaries=single_partial.boundaries,
+                            )
+                        ]
+                    )
+
+                next_step.set_result([RayMaterializedResult(partition, accessor, 0) for partition in next_step.inputs])
+                next_step.set_done()
+
+            # CASE: Actual task that needs to be dispatched
+            else:
+                tasks_to_dispatch.append(next_step)
+
+        return tasks_to_dispatch, True
+
+    def _dispatch_tasks(
+        self,
+        tasks_to_dispatch: list[PartitionTask],
+        daft_execution_config: PyDaftExecutionConfig,
+    ) -> Iterator[tuple[PartitionTask, list[ray.ObjectRef]]]:
+        """Iteratively Dispatches a batch of tasks to the Ray backend"""
+
+        for task in tasks_to_dispatch:
+            if task.actor_pool_id is None:
+                results = _build_partitions(daft_execution_config, task)
+            else:
+                actor_pool = self._actor_pools.get(task.actor_pool_id)
+                assert actor_pool is not None, "Ray actor pool must live for as long as the tasks."
+                results = _build_partitions_on_actor_pool(task, actor_pool)
+            logger.debug("%s -> %s", task, results)
+
+            yield task, results
+
+    def _await_tasks(
+        self,
+        inflight_ref_to_task_id: dict[ray.ObjectRef, str],
+    ) -> list[ray.ObjectRef]:
+        """Awaits for tasks to be completed. Returns tasks that are ready.
+
+        NOTE: This method blocks until at least 1 task is ready. Then it will return as many ready tasks as it can.
+        """
+        if len(inflight_ref_to_task_id) == 0:
+            return []
+
+        # Await on (any) task to be ready with an unlimited timeout
+        ray.wait(
+            list(inflight_ref_to_task_id.keys()),
+            num_returns=1,
+            timeout=None,
+            fetch_local=False,
+        )
```

Comment:
NOTE: slight behavior change here compared to previous awaiting logic

I call a `ray.wait` here but discard the outputs to just wait on one item to be ready with `timeout=None`.

Then I subsequently call `ray.wait` again with a `0.01` timeout to actually retrieve a batch of ready tasks.

I think this logic is a little easier to follow, and gets rid of the weird loop over `("next_one", "next_batch")` that we had earlier. Also shouldn't have too much of a performance impact.